File: 排序/mergeSort.js
let mergeSort=(arr)=>{
    let len=arr.length
    if(arr.length===1){
        return arr
    }else {
        let mid=Math.floor(len/2),
        left=arr.slice(0,mid),
        right=arr.slice(mid)
        return mergeSort(mergeSort(left),mergeSort(right))
    }
}
function merge (left,right){
    const tmp = [];
    while (left.length && right.length) {
      // 合并左右数组
      if (left[0] < right[0]) tmp.push(left.shift());
      else tmp.push(right.shift());
    }
    return tmp.concat(left, right);
}

File: 排序/mergeSort.go
/**
归并排序（Merge sort，台湾译作：合并排序）是建立在归并操作上的一种有效的排序算法。该算法是采用分治法（Divide and Conquer）的一个非常典型的应用
分治思想，时间复杂度为：O(n*log(n))
*/

func mergeSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}
	i := len(arr) / 2
	left := mergeSort(arr[0:i])
	right := mergeSort(arr[i:])
	result := merge(left, right)
	return result
}

func merge(left, right []int) []int {
	result := make([]int, 0)
	m, n := 0, 0 // left和right的index位置
	l, r := len(left), len(right)
	for m < l && n < r {
		if left[m] > right[n] {
			result = append(result, right[n])
			n++
			continue
		}
		result = append(result, left[m])
		m++
	}
	result = append(result, right[n:]...) // 这里竟然没有报数组越界的异常？
	result = append(result, left[m:]...)
	return result
}
